### PR TITLE
Fix CloudFront cache_behavior precedence

### LINKED ISSUE #7253: Allow configuring CloudFront cache_behavior precedence
### Terraform Version

0.6.16
### Affected Resource(s)

Please list the resources as a list, for example:
- cloudfront_distribution > cache_behavior
### Terraform Configuration Files

`````` hcl
# Distro pointing at the specified domain.
resource "aws_cloudfront_distribution" "cloudfront" {
  origin {
    domain_name = "${var.lde_vm_domain_name}"
    origin_id = "lde_vm"
    custom_origin_config {
      http_port = 80
      https_port = 443
      origin_protocol_policy = "match-viewer"
      origin_ssl_protocols = [
        "SSLv3",
        "TLSv1"]
    }
    custom_header = {
      name = "${var.lde_custom_header_name}"
      value = "${var.lde_custom_header_value}"
    }
  }
  aliases = ["${var.public_domain_name}"]
  enabled = true
  comment = "CloudFront Server"
  default_cache_behavior {
    target_origin_id = "lde_vm"
    allowed_methods = [
      "DELETE",
      "GET",
      "HEAD",
      "OPTIONS",
      "PATCH",
      "POST",
      "PUT"]
    cached_methods = [
      "GET",
      "HEAD"]
    compress = true
    smooth_streaming = false
    forwarded_values {
      cookies {
        forward = "all"
      }
      headers = ["Host"]
      query_string = true
    }
    viewer_protocol_policy = "allow-all"
    min_ttl = 0
    default_ttl = 3600
    max_ttl = 31536000
  }

# private digital content cache
  cache_behavior {
    path_pattern = "user/digital_content/download/*"
    target_origin_id = "lde_vm"
    allowed_methods = [
      "GET",
      "HEAD",
      "OPTIONS"]
    cached_methods = [
      "GET",
      "HEAD"]
    compress = true
    smooth_streaming = false
    forwarded_values {
      cookies {
        forward = "none"
      }
      headers = ["Host"]
      query_string = true
    }
    trusted_signers = ["self"]
    viewer_protocol_policy = "allow-all"
    min_ttl = 0
    default_ttl = 3600
    max_ttl = 31536000
  }

  # JS files
  cache_behavior {
    path_pattern = "js/*"
    target_origin_id = "lde_vm"
    allowed_methods = [
      "GET",
      "HEAD",
      "OPTIONS"]
    cached_methods = [
      "GET",
      "HEAD"]
    compress = true
    smooth_streaming = false
    forwarded_values {
      cookies {
        forward = "none"
      }
      headers = ["Host"]
      query_string = true
    }
    viewer_protocol_policy = "allow-all"
    min_ttl = 0
    default_ttl = 3600
    max_ttl = 31536000
  }

  # CSS files
  cache_behavior {
    path_pattern = "css/*"
    target_origin_id = "lde_vm"
    allowed_methods = [
      "GET",
      "HEAD",
      "OPTIONS"]
    cached_methods = [
      "GET",
      "HEAD"]
    compress = true
    smooth_streaming = false
    forwarded_values {
      cookies {
        forward = "none"
      }
      headers = ["Host"]
      query_string = true
    }
    viewer_protocol_policy = "allow-all"
    min_ttl = 0
    default_ttl = 3600
    max_ttl = 31536000
  }

  restrictions {
    geo_restriction {
      restriction_type = "none"
    }
  }

  viewer_certificate {
    iam_certificate_id = "${aws_iam_server_certificate.alias_certificate.id}"
    ssl_support_method = "sni-only"
    minimum_protocol_version = "TLSv1"
  }
  retain_on_delete = false
  price_class = "${var.cloudfront_price_class}"
}```

### Debug Output
N/A

### Panic Output
N/A

### Expected Behavior
We should be able to specify the precedence of multiple cache_behavior definitions. Eg:

```hcl
  cache_behavior {
    precedence: 0
    path_pattern: "js/*"
    ...
  }

  cache_behavior {
    precedence: 1
    path_pattern: "images/*"
  }
``````

Alternately, the order that the cache_behaviour blocks are defined in the config should define the precedence.
### Actual Behavior

Cache behaviours are created in random order:

aws_cloudfront_management_console
### Steps to Reproduce

Please list the steps required to reproduce the issue, for example:
1. `terraform apply`
### Important Factoids

Are there anything atypical about your accounts that we should know? For example: Running in EC2 Classic? Custom version of OpenStack? Tight ACLs?

Nothing special.
### References

Not that I'm aware of.

### PR DESCRIPTION
**Pull request by optimisticanshul**
Fixes #7253

### CODE REVIEW

**Comment by jen20** [comment]
Hi @optimisticanshul! Thanks for the pull request here. This looks correct to me, but I'm going to defer to @catsby on this, as I can't remember offhand whether there may be unintended side effects here.

**Comment by radeksimko** [comment]
Hi @optimisticanshul 
I wish the fix was that easy! 😺 
I did hit the same issue about 2 weeks ago and started working on a fix.

There's a few side effects and related issues yet to tackle

TL;DR:

1. [expand/flatten logic](https://github.com/hashicorp/terraform/blob/master/builtin/providers/aws/cloudfront_distribution_configuration_structure.go#L231) will need to be updated to reflect this new change (basically `m["field"].(*schema.Set).List()` needs to be changed to `m["field"].([]interface{})`)
2. unit tests for expand & flatten logic need to be updated to reflect this (e.g. `reflect.DeepEqual` doesn't play very well with nested `schema.Set`s as these are commonly represented as pointers to `Set`s)
3. it's not just `cache_behavior`, `default_cache_behavior` may also needs to be changed or the expand/flatten logic around it needs to be changed.
4. Any existing state will require a migration plan to prevent crashes or any strange behavior for users that created CF distribution prior to the bugfix. That migration logic also needs a test case.
5. _(possibly unrelated, but may be fixed in the same PR)_ `*_methods` on the other hand need the exact opposite change (`TypeList` -> `TypeSet`) as the API returns these in random order and that generates extra diffs
6. acceptance test for verifying the fixed functionality (changing precedence of `cache_behavior`) should be added
7. the core `helper/schema` package doesn't handle nested `TypeSet` in `TypeList` well. `(d = ResourceData) d.Get("nested_set")` inside Update function will return something like `[]interface{}{}`.

I'm quite confident I nearly fixed `1.-5.` in my WIP branch but spent most of my time in the past days/weeks digging into 7.

I will try to share a repro case tonight in a form of PR.

Sorry for bad news! 😢

**Comment by vancluever** [comment]
Hey guys,

Sorry this is causing so much grief! 😳 

@radeksimko if you put the PR up and need help finishing it off @ me when you do it and I will put in some time as well.

**Comment by tmatilai** [comment]
Hi @radeksimko, would be easier to fix this issue now after those other pull requests merged?
